### PR TITLE
config: support extrinsic yaml anchor declarations

### PR DESCRIPTION
**Pull request by goaway**
Commit Message: config: support extrinsic yaml anchor declarations
Additional Description: This relatively simple change adds the ability to declare yaml anchors outside the bounds of Envoy configuration structure, for later aliasing and merging, using a new `!ignore` tag. This allows for bootstrap and static configuration to include variable content without requiring secondary pre-parsing or interpolation, by leveraging the yaml parser cheaply and directly. In the case of Envoy Mobile, this dramatically improves startup latency, and it likely has applications elsewhere.
Risk Level: Moderate
Testing: Locally with Envoy Mobile, and Unit test in CI
Docs Changes: Configuration examples

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by goaway** [comment]
@snowp

**Comment by wrowe** [comment]
Is there an appropriate docs change you've identified so this is highlighted in the release? Seems rather buried.

You may want to merge main to pick up newer coverage thresholds and pass CI.

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks great to me! Maybe include a version note about this?

@phlax this change might be interesting to you if you want to take a look, maybe there are use cases for this in existing documentation/examples

File: docs/root/configuration/overview/examples.rst [hunk 1]
```diff
@@ -302,3 +302,64 @@ The management server could respond to EDS requests with:
             socket_address:
               address: 127.0.0.2
               port_value: 1234
+
+Special YAML usage
+~~~~~~~~~~~~~~~~~~
+
+When loading YAML configuration, the Envoy loader will interpret nodes tagged with !ignore specially,
+and omit them from them entirely from the native configuration tree. Ordinarily, the YAML stream
```
Comment (phlax) [inline docs/root/configuration/overview/examples.rst:310]: "...and omit them entirely from the..."

File: docs/root/configuration/overview/examples.rst [hunk 2]
```diff
@@ -302,3 +302,64 @@ The management server could respond to EDS requests with:
...
+declared that is explicitly handled as a non-represented type.
+
+One potential use for this mechanism is to avoid extra interpolation or pre-processing overhead when
+dynamically generating YAML configuration for Envoy. Dynamic YAML anchors may be provided at the head
+of the stream, followed by a "template" that aliases those values.
+
+See the following example:
+
+.. validated-code-block:: yaml
+  :type-name: envoy.config.bootstrap.v3.Bootstrap
```
Comment (phlax) [inline docs/root/configuration/overview/examples.rst:321]: could you use a `literalinclude` rather than `validated-code-block`

File: docs/root/configuration/overview/examples.rst [hunk 3]
```diff
@@ -302,3 +302,64 @@ The management server could respond to EDS requests with:
...
+
+  dynamic_sockets: !ignore
```
Comment (phlax) [inline docs/root/configuration/overview/examples.rst:323]: the issue with this afaict is that it means the config is not loadable with a standard yaml parser - at least trying to load it with pyyaml failed

```
yaml.constructor.ConstructorError: could not determine a constructor for the tag '!ignore'
  in "<unicode string>", line 1, column 18:
    dynamic_sockets: !ignore

```

this ticket https://github.com/yaml/pyyaml/issues/266 suggests it should be possible with a full yaml loader (ie not safe_load) but i couldnt seem to make it work

**Review by phlax — CHANGES_REQUESTED** [review]
thanks @goaway i defo like the idea but i had some issues with trying to load the config with the python parser

i would really appreciate if you could move this to a literalinclude as it means the config will be tested loading in Envoy

File: docs/root/configuration/overview/examples.rst (hunk 3 above)
Comment (goaway) [reply docs/root/configuration/overview/examples.rst:323]: I'm not a subject matter expert, but based on my read of https://yaml.org/spec/1.2/spec.html this a correct and adherent usage of an application-specific local tag. This tag informs Envoy's (application-specific) loader of the type of the deserialized representation (in this case "nothing"). In theory, a validator should not complain, because this is valid yaml. In practice, a given YAML loader might complain, based on how it's configured, because it believes it should be doing something with the tag, and hasn't been told what to do. From my own Googling, various YAML loading tools, including pyyaml, provide various mechanisms for informing the library of how to handle custom local tags.

See also 3.3.2 and 3.3.3:
https://yaml.org/spec/1.2/spec.html#id2769212

Especially:
> That said, tag resolution is specific to the application. YAML processors should therefore provide a mechanism allowing the application to override and expand these default tag resolution rules.

I haven't attempted, but this stack overflow post appears to explain how to configure pyyaml to handle a custom tag:
https://stackoverflow.com/questions/43058050/creating-custom-tag-in-pyyaml

**Comment by goaway** [comment]
Thank you all for the comments so far!

@snowp, what format would I use for adding the release notes?

File: docs/root/configuration/overview/examples.rst [hunk 4]
```diff
@@ -306,10 +306,10 @@ The management server could respond to EDS requests with:
 Special YAML usage
 ~~~~~~~~~~~~~~~~~~
 
-When loading YAML configuration, the Envoy loader will interpret nodes tagged with !ignore specially,
-and omit them from them entirely from the native configuration tree. Ordinarily, the YAML stream
-must adhere strictly to the proto schemas defined for Envoy configuration. This allows content to be
-declared that is explicitly handled as a non-represented type.
+When loading YAML configuration, the Envoy loader will interpret nodes tagged with !ignore
```
Comment (phlax) [inline docs/root/configuration/overview/examples.rst:309]: nit - use ` ``!ignore`` ` as this is a literal

File: docs/root/configuration/overview/examples.rst (hunk 3 above)
Comment (phlax) [reply docs/root/configuration/overview/examples.rst]: yep - i didnt try too hard to get it to work tbh and yaml parsers can a bit "ymmv" - defo not a blocker

**Comment by goaway** [comment]
Thank you @phlax!

File: docs/root/configuration/overview/examples.rst (hunk 3 above)
Comment (goaway) [reply docs/root/configuration/overview/examples.rst]: Actually, looks like it's going to be necessary to pass bazel-bin/configs/example_configs_validation - let me see what I can come up with. :)

**Comment by wrowe** [comment]
Merging main should get you past the format_pre breakage.

File: tools/config_validation/validate_fragment.py [hunk 5]
```diff
@@ -21,6 +21,28 @@
...
+    yaml_tag = u'!ignore'
+
+    def __init__(self, strval):
+       self.strval = strval
+
+    def __repr__(self):
+       return f'IgnoredKey({str})'
+
+    def __eq__(self, other):
+        return isinstance(other, IgnoredKey) && self.strval == other.strval
```
Comment (phlax) [inline tools/config_validation/validate_fragment.py:35]: this isnt valid python - it needs to be `and` rather than `&&`

File: tools/config_validation/validate_fragment.py [hunk 6]
```diff
@@ -21,6 +21,28 @@
 import argparse
 
 
+class IgnoredKey(yaml.YAMLObject):
+    """Python support type for Envoy's config !ignore tag."""
+    yaml_tag = u'!ignore'
```
Comment (phlax) [inline tools/config_validation/validate_fragment.py:26]: no need for the `u` prefix in python 3

File: docs/root/configuration/overview/_include/tagged.yaml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+:type-name: envoy.config.bootstrap.v3.Bootstrap
```
Comment (phlax) [inline docs/root/configuration/overview/_include/tagged.yaml:1]: this shouldnt be here

**Comment by phlax** [comment]
im wondering whether the approach taken by docker-compose might be useful here - rather than adding `!ignore`

in docker-compose there are "extension fields" - ie fields prefixed with `x-`.

These are ignored by the parser and this allows you to use them for definitions https://docs.docker.com/compose/compose-file/compose-file-v3/#extension-fields

**Comment by goaway** [comment]
@phlax thanks, I just pushed to let CI check the python for me rather than figure out how to run it locally. I'm not opposed to the docker "magic prefix" approach, and even considered it - just went with this because it seemed maybe less likely to conflict with anything.

The prefix approach _may_ be even more work to support in python though - I think I'll have to manually traverse the dictionary tree to remove all prefixed keys, since the "skip keys" approach won't work.

**Comment by phlax** [comment]
> The prefix approach may be even more work to support in python though - I think I'll have to manually traverse the dictionary tree to remove all prefixed keys, since the "skip keys" approach won't work.

yep - perhaps in terms of the envoy parser - but it will make it more standard yaml and most likely easier to parse

it also feels a bit less magic than `!ignore` 

i have one eye here on adding vscode/jsonschema validation - and i think for that the `x-foo` approach is going to be a lot easier going than `!ignore`

**Comment by goaway** [comment]
Other than the manual work, my concern about a magic prefix was it will make it so that otherwise valid proto schema, including extensions, that happen to have a field that starts with a big X (or whatever the prefix is) will no longer be valid. That seems like that's arguably a backwards incompatible change and maybe more of a gotcha than a tag that only need see use in systems who deliberately choose to deploy it.

In other words, a given Envoy deploymenit has to *opt in* to be affected by the new tag, but every Envoy deployment has the potential to be affected by the prefix.

**Comment by goaway** [comment]
FWIW I don't personally have strong feelings one way or the other, but that's what tipped me over to the side of exploring a tag-based solution initially.

**Comment by phlax** [comment]
> ...field that starts with a big X (or whatever the prefix is) will no longer be valid. That seems like that's arguably a backwards incompatible change and maybe more of a gotcha than a tag that only need see use in systems who deliberately choose to deploy it

fair point - but im not seeing any existing *keys* that use this notation - only values

doing

```
$ git grep "x-" | grep yaml
```

seems at least to suggest that there are no keys that would be affected

cc @htuch for some overview here

**Comment by phlax** [comment]
> I think I'll have to manually traverse the dictionary tree to remove all prefixed keys, since the "skip keys" approach won't work.

iiuc this point - i dont think so - i think we should only allow definitions at the top level

**Comment by alyssawilk** [comment]
also cc @adisuissa

**Comment by goaway** [comment]
I actually figured out how to run the python validator locally, so this should work now.

**Comment by goaway** [comment]
Having slept on this, I'm feeling somewhat more inclined to prefer a tag-based approach over a magic-prefix one, largely for the reasons I mention above, and because compliant yaml parsers should know about tags, but I do acknowledge that there's precedent with Docker using prefixes, and that it may be somewhat more straightforward to handle prefixes with some toolchains.

File: docs/root/configuration/overview/examples.rst [hunk 8]
```diff
@@ -302,3 +302,20 @@ The management server could respond to EDS requests with:
...
+~~~~~~~~~~~~~~~~~~
+
+When loading YAML configuration, the Envoy loader will interpret map keys tagged with !ignore
+specially, and omit them entirely from the native configuration tree. Ordinarily, the YAML stream
+must adhere strictly to the proto schemas defined for Envoy configuration. This allows content to
+be declared that is explicitly handled as a non-represented type.
+
+One potential use for this mechanism is to avoid extra interpolation or pre-processing overhead when
+dynamically generating YAML configuration for Envoy. Dynamic YAML anchors may be provided at the head
+of the stream, followed by a "template" that aliases those values.
```
Comment (htuch) [inline docs/root/configuration/overview/examples.rst:316]: This is a bit confusing, can you give an example of how it is supposed to work? It seems you want some kind of templating, but unclear why parameters need to be ignored to do this.

Comment (goaway) [reply docs/root/configuration/overview/examples.rst:316]: @htuch There's an example in the attached yaml file. I hoped it would be illustrative of the intent here.

The reason parameters need to be ignored is that otherwise YAML anchors must be declared in-line to the structure of the document. (For instance, if you wanted to template some TLS parameters, your first usage would have to be literally declared in the first cluster, and subsequent clusters could reference it.) By allowing nodes to be defined that will be excluded from the deserialized config structure, we can separate a "pure" template containing only aliases, from declarations of dynamic content, containing only anchors. In both the attached example, and in Envoy Mobile, we intend to leverage this by passing yaml effectively in two chunks - one that's ignored by deserialization and contains dynamic values, and one that contains a fixed template. By separating them, we can avoid an extra interpolation or preprocessing step.

By way of another example, I've seen deployments rely on templating tools like jinja2 to template static/bootstrap config in the past. With this setup, they could avoid the extra tool and preprocessing step, and simply output appropriately serialized dynamic content followed by their template -- the loader will handle the rest.

Please let me know if that's still unclear, or if you have thoughts on how I can improve upon the description or example.

**Comment by phlax** [comment]
> I'm feeling somewhat more inclined to prefer a tag-based approach over a magic-prefix one

i think a way that works out of the box with standard parsers without having to code custom classes etc has a pretty big advantage

it makes the config files readable and writeable easily without having to write a load of additional code

**Comment by phlax** [comment]
also - i havent tested yet - but im concerned that getting this working with jsonschema and yaml-language-server - as required to get vscode auto-complete and validation working is going to be more complex if it works at all

**Comment by goaway** [comment]
@phlax I appreciate your concerns. Tags have been a part of the yaml standard since 1.0, and implicitly every node has one. Compliant yaml loaders really _must_ be able to handle them (see, e.g. https://yaml.org/spec/1.0/#id2490971).

FWIW, this approach worked out of the box with yaml-cpp, and worked just fine with pyyaml as well. The python `safe_load` requires the an explicit constructor for deserializing an application tag just because that's what makes it safe. Honestly, I probably could have skipped the custom class and just given the loader the string constructor, and that would have likely worked, too - the extra type just gave me a trivial path (`skipkeys`) to exclude it (as intended) from later processing.

My main concern with the magic prefix remains that it could break private extensions that other deployments have. It's a "gotcha" that existing deployments can't opt out of, whereas when it comes to this tag, any deployment that doesn't opt in to this feature will remain unaffected by this change.

**Comment by goaway** [comment]
All that said, this is not the hill I'm prepared to die on. ;)

If there's a strong consensus that a magic prefix would be a preferable mechanism, I'll just reimplement. At the end of the day it's the functionality that I'm really after, not a singular implementation.

**Comment by mattklein123** [comment]
My personal preference is to go with the YAML standard vs. the magic prefix approach, since the latter is by definition "magical" and is more likely to break someone.

@phlax this patch solves a very real problem for Envoy Mobile so my preference would be to merge it. If we are concerned about user confusion I think we could talk about leaving it undocumented, though my preference would still be to document it since it seems like it's a feature of the YAML standard.

File: docs/root/configuration/overview/examples.rst (hunk 8 above)
Comment (htuch) [reply docs/root/configuration/overview/examples.rst:316]: Keeping things simple, which is always good for ESL etc, I would just say "This lets you split your file into two parts: one in which we have YAML content not subject to parsing according to the schema and another part that is parsed. YAML anchors in the first part may be referenced by the second part."

Or something very simple like that. It honestly took me some time to parse "
 extra interpolation or pre-processing overhead whendynamically generating YAML configuration for Envoy". Like, "extra" compared to what? Who is interpolating where? What is the preprocessor and how?

Sorry to be picky, I'm just trying to feedback to better docs here.

Comment (goaway) [reply docs/root/configuration/overview/examples.rst:316]: No worries, I never feel very certain that I'm explaining things clearly, so this is helpful feedback. I'll take another crack at the wording with your suggestion. Thank you!

Comment (goaway) [reply docs/root/configuration/overview/examples.rst]: @htuch I incorporated your suggestion pretty directly. Please let me know if that seems clearer.

**Review by mattklein123 — APPROVED** [review]
This LGTM but will defer to @htuch and @phlax on any further comments.

**Review by htuch — DISMISSED** [review]
LGTM, thanks!

**Comment by goaway** [comment]
I think the CI failures might be spurious - this was green before the docs change. Not sure if I can /retest

**Comment by phlax** [comment]
> My personal preference is to go with the YAML standard vs. the magic prefix approach, since the latter is by definition "magical" and is more likely to break someone. 

i think there is some confusion here

the `!ignore ` syntax is much more likely to break someone - it may be part of the standard - but it is not standard

the `x-definitions` way is "standard" yaml - there isnt even any need for the `x-` part we could just make it `definitions` - so there is not much magical here - its just ignored on the basis of having an `x-` prefix much like http headers are.

Prefixing with `!ignore ` is no more/less magical than prefixing with `x-` 

Using `x-` to denote extension fields is more standard broadly - eg that is what is used in docker-compose - but you could also point to the http header example for that matter.

>  this patch solves a very real problem for Envoy Mobile so my preference would be to merge it.

i appreciate that and welcome the addition of this - but its a significant change and should be thought about properly and not rushed

>  If we are concerned about user confusion I think we could talk about leaving it undocumented, though my preference would still be to document it since it seems like it's a feature of the YAML standard.

my concerns have nothing to do with docs

broadly my concerns are as follows

- its not supported out-of-the box by parsers - at least not by the python parser and certainly not in safe mode (which you want as yaml also allows for embedded external code execution)
- that it wont work with yaml-language-server and therefore wont be usable once we have added vscode support

for clarity - here are 2 examples showing the different approaches

```
!ignore dynamic_sockets:
  &admin_address
  address: 127.0.0.1
  port_value: 9901

admin:
  address:
    socket_address: *admin_address
```

```
x-dynamic_sockets:
  &admin_address
  address: 127.0.0.1
  port_value: 9901

admin:
  address:
    socket_address: *admin_address
```

i have tried again parsing using both `load` and `safe_load` -  afaict hacking `SafeLoader` and `SafeDumper` is required so this pushes more work onto anyone wanting to parse config files (at least if they support this syntax)

wrt using this with yaml-language-server - i have checked and wasnt able to find any evidence that the `!ignore` syntax is supported (or to the contrary) - the `x-format` will work inherently

there is still a further challenge of expressing this in jsonschema to make it work with yaml-language-server with either way

overall, im -1 to the approach implemented here because:

- its **more** likely to break end users parsing of config
- it pushes the implementation of a custom parser on to end users - i think unnecessarily
- extension headers with a `x-` prefix is more standard (eg docker-compose using it)
- its much less likely to work with vscode or other editors

**Comment by phlax** [comment]
> We may differ on our notion of "out-of-the-box", 

my intended meaning of `out-of-the-box` is that there is no need for any custom classes

ie `yaml.safe_load(content)` just works

> but we are making use of an official public interface as intended,

yes - but with a custom class with a fair amount of complexity to do so

**Comment by phlax** [comment]
...also without reading the Envoy source code - there is not any obvious way for an end user to know *how* to do this

**Comment by mattklein123** [comment]
OK thanks @phlax I think I understand the issue better now. I didn't realize that `!ignore` is magic either here. Here is my suggestion:
1) Pick whatever magic string (`x-` prefix or whatever) we think is most standard in the industry.
2) Clearly document/warn that use of this feature may break generic parsing of the YAML and/or require special code at the parsing site. Potentially also link to the internal Python changes in the validator to wire this up.

I think that would be a reasonable compromise to move this forward.

The other alternative would be to just not document it at all, though that would not be my first choice.

**Comment by mattklein123** [comment]
OK sorry for the back and forth on this. I just discussed this further with @goaway and I better understand this issue.

I agree that using tags vs. magic strings is a better way of handling this. As such my opinion is we should stick with the `!ignore` approach. Either way the YAML loader if verifiers, etc. will need to be modified for whatever we choose.

So concretely I think we should:
1) Keep the code change
2) Add a warning in the documentation that discusses that use of this feature may break projects that load Envoy YAML without further modification, and also link to the example changes in the Python validation code in the project. I think this is sufficient to move forward.

**Review by mattklein123 — APPROVED** [review]
Awesome, thanks. @phlax let's chat early in the morning tomorrow on Slack and resolve any remaining issues so we can move this forward. It's fairly urgent we get this merged so if we can't agree to this quickly we will need to carry a patch in Envoy Mobile.

**Comment by mattklein123** [comment]
docs needs fixing

File: docs/root/configuration/overview/examples.rst (hunk 8 above)
Comment (htuch) [reply docs/root/configuration/overview/examples.rst]: LGTM

File: docs/root/configuration/overview/examples.rst [hunk 9]
```diff
@@ -328,9 +328,10 @@ See the following example:
 
 For example, this will instruct `PyYAML <https://github.com/yaml/pyyaml>` to treat an ignored
 node as a simple scalar when loading:
-.. code-block:: python
+
+.. code-block:: python3
```
Comment (phlax) [inline docs/root/configuration/overview/examples.rst:332]: not sure this is any different from `python` lexer - i think if not just use `python` as that is what is used everywhere else

**Comment by phlax** [comment]
just to confirm the `!ignore` syntax wont work with yaml-language-server 

it throws a couple of errors

it might be possible to edit vscodes' `settings.json` and workaround - but the point is we just want to distribute `envoy.schema.json` and not have to tell people to add additional settings beyond registering the schema

**Comment by phlax** [comment]
a fixed key such as `definitions` would obv work out-of-the-box at least as far as validation goes

the `x-syntax` could also be expressed with pattern properties - https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties

**Comment by mattklein123** [comment]
@phlax objection noted but I think it's time to move this forward. I think it's fine if this feature does not work with the language server. That's why we have the warning. We can ship and iterate on this later if needed. (@goaway please do a follow up for the docs comments but I don't think we need to wait for more CI on this.)

**Review by mattklein123 — APPROVED** [review]
